### PR TITLE
Adding EqualsTester

### PR DESCRIPTION
**Pull request by adambender**

### CODE REVIEW

File: lib/testing/src/equality/equality.dart [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+
+/**
+ * Matcher for == and hashCode methods of a class.
+ *
+ * To use, invoke areEqualityGroups with a list of equality groups where each
+ * group contains objects that are supposed to be equal to each other, and
+ * objects of different groups are expected to be unequal. For example:
+ *
+ * expect({
```
Comment (yjbanov) [inline lib/testing/src/equality/equality.dart]: This line needs 4-space indent to render properly in dartdocs

File: lib/testing/src/equality/equality.dart [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+ *      'hello': ["hello", "h" + "ello"],
+ *      'world': ["world", "wor" + "ld"],
+ *      'three': [2, 1 + 1]
+ *     }, areEqualityGroups);
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: Question. Which do people prefer: `areEqualityGroups` or `areGroupwiseEqual`?

File: lib/testing/src/equality/equality.dart [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+ *
+ * This tests:
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: This tests that:

File: lib/testing/src/equality/equality.dart [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+ *     true
+ * * comparing each pair of objects from different equality groups returns
+ *     false
+ * * the hash codes of any two equal objects are equal
+ *
+ *
+ * When a test fails, the error message labels the objects involved in
+ * the failed comparison as follows:
+ *
+ *      "`[group i, item j]`" refers to the jth item in the ith equality group,
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: "`[group x, item i]`" refers to the ith item in the equality group named x.

File: lib/testing/src/equality/equality.dart [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+ * This tests:
+ *
+ * * comparing each object against itself returns true
+ * * comparing each object against an instance of an incompatible class
+ *     returns false
+ * * comparing each pair of objects within the same equality group returns
+ *     true
+ * * comparing each pair of objects from different equality groups returns
+ *     false
+ * * the hash codes of any two equal objects are equal
```
Comment (yjbanov) [inline lib/testing/src/equality/equality.dart]: It also tests determinism of the equality functions

File: lib/testing/src/equality/equality.dart [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+ * When a test fails, the error message labels the objects involved in
+ * the failed comparison as follows:
+ *
+ *      "`[group i, item j]`" refers to the jth item in the ith equality group,
+ *       where both equality groups and the items within equality groups are
+ *       numbered starting from 1.  When either a constructor argument or an
+ *       equal object is provided, that becomes group 1.
+ *
+ */
+const REPETITIONS = 3;
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: Here and below: constants in camelCase.

File: lib/testing/src/equality/equality.dart [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+const NO_MISMATCH_FOUND = 'therewasnomismatchfoundhere';
```
Comment (yjbanov) [inline lib/testing/src/equality/equality.dart]: Is this used anywhere?

File: lib/testing/src/equality/equality.dart [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+ *
+ * When a test fails, the error message labels the objects involved in
+ * the failed comparison as follows:
+ *
+ *      "`[group i, item j]`" refers to the jth item in the ith equality group,
+ *       where both equality groups and the items within equality groups are
+ *       numbered starting from 1.  When either a constructor argument or an
+ *       equal object is provided, that becomes group 1.
+ *
+ */
```
Comment (yjbanov) [inline lib/testing/src/equality/equality.dart]: This doc comment should be applied to `areEqualityGroups`, not `REPETITIONS`.

File: lib/testing/src/equality/equality.dart (hunk 6 above)
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: Make private.

File: lib/testing/src/equality/equality.dart (hunk 7 above)
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: This doesn't appear to be used anywhere.

File: lib/testing/src/equality/equality.dart [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        .add(" ${matchState[FAILURE_REASON]}");
+
+
+  void _verifyEqualityGroups(Map<String, List<Object>> equalityGroups,
+      Map matchState) {
+    if (equalityGroups == null) {
+      throw new MatchError('Equality Group must not be null');
+    }
+    var equalityGroupsCopy = {};
+    equalityGroups.keys.forEach((String groupName) {
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: Consider `equalityGroups.forEach((String groupName, List<Object> group) {`

File: lib/testing/src/equality/equality.dart [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+      if (groupName == null) {
+        throw new MatchError('Group name must not be null');
+      }
+      var group = equalityGroups[groupName];
+      if (group == null) {
+        throw new MatchError('Group must not be null');
+      }
+      equalityGroupsCopy[groupName] = new List.from(group);
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: Consider either using `List` everywhere, or use `List<Object>.from(group)` here for consistency.

File: lib/testing/src/equality/equality.dart [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+      }
+
+      if (item != item) {
+        throw new MatchError("$item must be Object#equals to itself");
+      }
+
+      var a = item.hashCode;
+      var b = item.hashCode;
+      if (a != b) {
+        throw new MatchError("the Object#hashCode of $item must be consistent");
```
Comment (yjbanov) [inline lib/testing/src/equality/equality.dart]: Might "idempotent" be more precise than "consistent"?

Comment (yjbanov) [inline lib/testing/src/equality/equality.dart]: Also, I'd say "the implementation of Object#hashCode ..."

File: lib/testing/src/equality/equality.dart [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+      }
+    }
+  }
+
+  void _checkGroupBasedEquality(Map<String, List<Object>> equalityGroups) {
+    equalityGroups.keys.forEach((String groupName) {
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: `equalityGroups.forEach((String groupName, List<Object> group) {`

File: lib/testing/src/equality/equality.dart [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+        }
+      }
+    });
+  }
+
+  void _checkEqualAgainstSameGroup(Map<String, List<Object>> equalityGroups,
+    int groupLength, int itemNumber, String groupName) {
+    for (var relatedItemNumber = 0; relatedItemNumber < groupLength;
+        relatedItemNumber++) {
+      if (itemNumber != relatedItemNumber) {
```
Comment (yjbanov) [inline lib/testing/src/equality/equality.dart]: I think the implementation might be simpler if checking within the same group included checking self equality. If equality check fails, then check the index to issue different error messages.

File: lib/testing/src/equality/equality.dart [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+            "object of another class");
+      }
+
+      if (item != item) {
+        throw new MatchError("$item must be Object#equals to itself");
+      }
+
+      var a = item.hashCode;
+      var b = item.hashCode;
+      if (a != b) {
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: Consider eliminating the locals; they don't add much in terms of readability.

File: lib/testing/src/equality/equality.dart (hunk 11 above)
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: Agreed on "idempotent"; "consistent" isn't very precise.

File: lib/testing/src/equality/equality.dart [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+        _checkEqualAgainstSameGroup(equalityGroups, groupLength, itemNumber,
+            groupName);
+        _checkUnequalsAgainstOtherGroups(equalityGroups, groupName, itemNumber);
+      }
+    });
+  }
+
+  void _checkUnequalsAgainstOtherGroups(Map<String,
+      List<Object>> equalityGroups, String groupName, int itemNumber) {
+    equalityGroups.keys.forEach((String unrelatedGroupName) {
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: `equalityGroups.forEach((String unrelatedGroupName, List<Object> unrelatedGroup) {`

File: lib/testing/src/equality/equality.dart [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+
+part of quiver.testing.equality;
+
+
+/**
+ * Matcher for == and hashCode methods of a class.
+ *
+ * To use, invoke areEqualityGroups with a list of equality groups where each
+ * group contains objects that are supposed to be equal to each other, and
+ * objects of different groups are expected to be unequal. For example:
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: Please document the format of the map to make it obvious that the keys are group names / serve only as documentation.

File: lib/testing/src/equality/equality.dart [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+      var groupLength = equalityGroups[groupName].length;
+      for (var itemNumber = 0; itemNumber < groupLength; itemNumber++) {
+        _checkEqualAgainstSameGroup(equalityGroups, groupLength, itemNumber,
+            groupName);
+        _checkUnequalsAgainstOtherGroups(equalityGroups, groupName, itemNumber);
+      }
+    });
+  }
+
+  void _checkUnequalsAgainstOtherGroups(Map<String,
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: Consider `_checkEqualToOtherGroup` and `_checkUnequalToOtherGroups`?

File: test/testing/equality/equality_test.dart [hunk 18]
```diff
@@ -0,0 +1,360 @@
...
+        expect({
+            'valid groups one': [named('foo').addPeers(['bar']),
+                             named('bar').addPeers(['foo'])],
+            'valid groups two': [named('baz'), named('baz')]
+          }, areEqualityGroups);
+      });
+  });
+}
+
+/// Test class that violates reflexitivity.  It is not equal to itself
```
Comment (cbracken) [inline test/testing/equality/equality_test.dart]: End with a period.

File: lib/testing/src/equality/equality.dart [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+
+class _Item {
+  final Object value;
+  final String groupName;
+  final int itemNumber;
+
+  _Item(this.value, this.groupName, this.itemNumber);
+
+  @override
+  String toString() => "$value [group '$groupName', item ${(itemNumber + 1)}]";
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: ... `item ${itemNumber + 1}]"`

File: lib/testing/src/equality/equality.dart [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+      _checkGroupBasedEquality(equalityGroupsCopy);
+    }
+  }
+
+  void _checkBasicIdentity(Map<String, List<Object>> equalityGroups,
+      Map matchState) {
+    var flattened = equalityGroups.values.expand((group) => group);
+    for (var item in flattened) {
+      if (item == _NotAnInstance.EQUAL_TO_NOTHING) {
+        throw new MatchError("$item must not be Object#equals to an arbitrary "
```
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: Replace uses of `Object#equals` with just "equal[s]" as the case warrants. It's verbose, and Dart Objects don't define an `equals` method, they define `operator ==` which is even more verbose.

File: lib/testing/src/equality/equality.dart (hunk 11 above)
Comment (cbracken) [inline lib/testing/src/equality/equality.dart]: On Object#hashCode, I'd just say "the implementation of hashCode".

File: lib/testing/src/equality/equality.dart (hunk 1 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 3 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 4 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 5 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 6 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 7 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 8 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 9 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 10 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Went with List, it was cleaner.

File: lib/testing/src/equality/equality.dart (hunk 11 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: done.

File: lib/testing/src/equality/equality.dart (hunk 14 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 15 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 16 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 2 above)
Comment (yjbanov) [inline lib/testing/src/equality/equality.dart]: I feel `areEqualityGroups` is more precise. Ok with keeping it. In fact I'm thinking we could be even more precise, e.g. `distinctEqualityGroups` to underscore non-equality across groups.

File: lib/testing/src/equality/equality.dart (hunk 17 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: test/testing/equality/equality_test.dart (hunk 18 above)
Comment (adambender) [inline test/testing/equality/equality_test.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 19 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 20 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 12 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

File: lib/testing/src/equality/equality.dart (hunk 2 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: I like the first myself.

Comment (yjbanov) [inline lib/testing/src/equality/equality.dart]: Chris also "kinda like[s]" `formDistinctEqualityGroups`

File: lib/testing/src/equality/equality.dart (hunk 13 above)
Comment (adambender) [inline lib/testing/src/equality/equality.dart]: Done.

**Comment by cbracken** [comment]
lgtm. @yjbanov, any other comments?

**Comment by cbracken** [comment]
Closed by 25497260deb6e33eab1e7deffaf2aa927e2d4765.

**Pull request closed without merging**